patches/server/1044-Add-Llama-caravan-events.patch

```diff
@@ -0,0 +1,98 @@
+             } else if (!llama.isLeashed() && !this.firstIsLeashed(llama, 1)) {
+                 return false;
+             } else {
+-                this.llama.joinCaravan(llama);
++                this.llama.joinCaravan(llama, io.papermc.paper.event.entity.LlamaJoinCaravanEvent.Reason.AI_GOAL); // Paper - add reason
+                 return true;
+             }
+         } else {
+@@ -97,7 +97,7 @@ public class LlamaFollowCaravanGoal extends Goal {
+ 
+     @Override
+     public void stop() {
+-        this.llama.leaveCaravan();
++        this.llama.leaveCaravan(io.papermc.paper.event.entity.LlamaLeaveCaravanEvent.Reason.LOST_HEAD); // Paper - add reason
+         this.speedModifier = 2.1D;
+     }
+ 
+diff --git a/src/main/java/net/minecraft/world/entity/animal/horse/Llama.java b/src/main/java/net/minecraft/world/entity/animal/horse/Llama.java
+index 5f61c97478f005aaaaad1b027118079db7275cf7..83ffa6b121460d769245c82337d06332e89a7289 100644
+--- a/src/main/java/net/minecraft/world/entity/animal/horse/Llama.java
++++ b/src/main/java/net/minecraft/world/entity/animal/horse/Llama.java
+@@ -437,8 +437,19 @@ public class Llama extends AbstractChestedHorse implements VariantHolder<Llama.V
+         }
+     }
+ 
++    // Paper start - Add Llama caravan events
+     public void leaveCaravan() {
++        leaveCaravan(io.papermc.paper.event.entity.LlamaLeaveCaravanEvent.Reason.UNKNOWN);
++    }
++    // Paper end - Add Llama caravan events
++
++    public void leaveCaravan(io.papermc.paper.event.entity.LlamaLeaveCaravanEvent.Reason reason) { // Paper - add reason
+         if (this.caravanHead != null) {
++            // Paper start - Add Llama caravan events
++            if (!new io.papermc.paper.event.entity.LlamaLeaveCaravanEvent(getBukkitEntity(), reason).callEvent()) {
++                return;
++            }
++            // Paper end - Add Llama caravan events
+             this.caravanHead.caravanTail = null;
+         }
+ 
+@@ -446,10 +457,26 @@ public class Llama extends AbstractChestedHorse implements VariantHolder<Llama.V
+     }
+ 
+     public void joinCaravan(Llama llama) {
++        joinCaravan(llama, io.papermc.paper.event.entity.LlamaJoinCaravanEvent.Reason.UNKNOWN);
++    }
++
++    public void joinCaravan(Llama llama, io.papermc.paper.event.entity.LlamaJoinCaravanEvent.Reason reason) { // Paper - add reason
++        // Paper start - Add Llama caravan events
++        if (!new io.papermc.paper.event.entity.LlamaJoinCaravanEvent(getBukkitEntity(), llama.getBukkitEntity(), reason).callEvent()) {
++            return;
++        }
++        // Paper end - Add Llama caravan events
+         this.caravanHead = llama;
+         this.caravanHead.caravanTail = this;
+     }
+ 
++    // Paper start
++    @Override
++    public org.bukkit.craftbukkit.entity.CraftLlama getBukkitEntity() {
++        return (org.bukkit.craftbukkit.entity.CraftLlama) super.getBukkitEntity();
++    }
++    // Paper end
++
```

Comment:
Not sure this is worth not just having 2 casts when the llama is passed to the event constructor. I don't think we do this for other similar situations.

Reply:
I dont really have a strong opinion on this. In our CraftLlama we could also remove some casts and for plugin devs it is always tedious to use  `((X) ((CraftX) x).getHandle()).methodCall()`
<img width="1682" alt="image" src="https://github.com/PaperMC/Paper/assets/31861387/ef3949a0-4a4e-4e4b-83c9-f9b295c72d8a">
We can also use casts but even with var it looks quite ugly using the fqn. Do you have fears of merge conflicts with upstream?